PYIC-5600: Modify datastructures

MikeCollingwood opened the pull request:
## Proposed changes

### What changed

- Added for loops to Nunjucks
- Added PAGES enum constant
- Used path library for path-wrangling

Use path whenever possible, but if there are queries to add: use URL library, and if the path has no host defined, use URI.

### Why did it change

- DRY
- simplicity
- consistency

### Issue tracking

- [PYIC-5600](https://govukverify.atlassian.net/browse/PYIC-5600)

### Notes

- `safe` is a built-in Nunjucks filter: [safe](https://mozilla.github.io/nunjucks/templating.html#safe), [autoescaping](https://mozilla.github.io/nunjucks/templating.html#autoescaping)
_"Mark the value as safe which means that in an environment with automatic escaping enabled this variable will not be escaped."_
NB: None of the content edited is escapable

[PYIC-5600]: https://govukverify.atlassian.net/browse/PYIC-5600?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

<img width="538" alt="Screenshot 2024-04-12 at 09 10 32" src="https://github.com/govuk-one-login/ipv-core-front/assets/144116535/92dd91bf-4336-485f-b1d5-6d6d92bbc751">
<img width="603" alt="Screenshot 2024-04-12 at 09 10 25" src="https://github.com/govuk-one-login/ipv-core-front/assets/144116535/06101306-0496-448b-9c7b-a32dd6f88d70">

MikeCollingwood committed:
PYIC-5600: Modify templates to use for loops

MikeCollingwood committed:
PYIC-5600: Use path and URL for uris and add ipv pages constant

blakeyp commented on src/app/credential-issuer/middleware.js:
```diff
@@ -50,7 +60,7 @@ module.exports = {
       authorizationCode: req.query?.code,
       credentialIssuerId: criId,
-      redirectUri: `${EXTERNAL_WEBSITE_HOST}/credential-issuer/callback/${criId}`,
+      redirectUri: callbackPath,
```
is this missing `criId` on the end? (and should be `.href`?)

MikeCollingwood replied:
It shouldn't need `.href` since callbackPath is a string rather than URL type. Good point on the `criId`!

blakeyp commented on src/lib/locals.js:
```diff
@@ -20,8 +21,15 @@ module.exports = {
+      "fromUrl",
+      `${SERVICE_URL}${req.originalUrl}`,
+    );
+    res.locals.contactUsUrl = contactUsUrl;
```
Does this need converting to a string?

blakeyp commented on src/lib/paths.js:
```diff
@@ -0,0 +1,24 @@
+}
+
+module.exports = {
+  getRoutePath: getPostEndpoint,
```
any reason not to just call the function getRoutePath?

blakeyp reviewed (commented):
Nice 🚀

MikeCollingwood committed:
PYIC-5600: Respond to PR comments:
- Fix missing criId dynamic redirectUri
- Fix conversion to string of contactUsUrl
- Refactor

MikeCollingwood commented on src/app/development/router.js:
```diff
@@ -7,13 +7,14 @@
-router.post("/all-templates", csrfProtection, allTemplatesPost);
-router.get("/all-templates", csrfProtection, allTemplatesGet);
+router.post(path.join("/", "all-templates"), csrfProtection, allTemplatesPost);
+router.get(path.join("/", "all-templates"), csrfProtection, allTemplatesGet);
```
@blakeyp Do you know if these CodeQL errors can be solved?

blakeyp replied:
I think we can safely ignore this, rate limiting for us should be done at the infra level not in code

MikeCollingwood committed:
PYIC-5600: Modify URI manipulation

MikeCollingwood committed:
PYIC-5600: Fix for rebase

MikeCollingwood committed:
PYIC-5624: Fix linting error
